primitives/phragmen/src/reduce.rs

```diff
@@ -0,0 +1,1065 @@
+							// valid because the two chained sections of `cycle` are aligned;
+							// index [0, 2] are both voting for v1 or both v2. Same goes for [1, 3].
+							decrease_indices.push(sibling_index - 2);
+							increase_indices.push(min_index - 2);
+						}
+
+						// apply changes
+						let mut remove_indices: Vec<usize> = Vec::with_capacity(1);
+						increase_indices.into_iter().for_each(|i| {
+							let voter = if i < 2 {
+								who.clone()
+							} else {
+								other_who.clone()
+							};
+							// Note: so this is pretty ambiguous. We should only look for one
+							// assignment that meets this criteria and if we find multiple then that
+							// is a corrupt input. Same goes for the next block.
+							assignments
+								.iter_mut()
+								.filter(|a| a.who == voter)
+								.for_each(|ass| {
+									ass.distribution
+										.iter_mut()
+										.position(|(t, _)| *t == cycle[i].0)
+										.map(|idx| {
+											let next_value =
+												ass.distribution[idx].1.saturating_add(min_value);
+											ass.distribution[idx].1 = next_value;
+										});
+								});
+						});
+						decrease_indices.into_iter().for_each(|i| {
+							let voter = if i < 2 {
+								who.clone()
+							} else {
+								other_who.clone()
+							};
+							assignments
+								.iter_mut()
+								.filter(|a| a.who == voter)
+								.for_each(|ass| {
+									ass.distribution
+										.iter_mut()
+										.position(|(t, _)| *t == cycle[i].0)
+										.map(|idx| {
+											let next_value =
+												ass.distribution[idx].1.saturating_sub(min_value);
+											if next_value.is_zero() {
+												ass.distribution.remove(idx);
+												remove_indices.push(i);
+												num_changed += 1;
+											} else {
+												ass.distribution[idx].1 = next_value;
+											}
+										});
+								});
+						});
+
+						// remove either one of them.
+						let who_removed = remove_indices.iter().find(|i| **i < 2usize).is_some();
+						let other_removed =
+							remove_indices.into_iter().find(|i| *i >= 2usize).is_some();
+
+						match (who_removed, other_removed) {
+							(false, true) => {
+								*other_who = who.clone();
+							}
+							(true, false) => {} // nothing, other_who can stay there.
+							(true, true) => {
+								entry.remove();
+							} // remove and don't replace
+							(false, false) => {
+								// Neither of the edges was removed? impossible.
+								debug_assert!(false, "Duplicate voter (or other corrupt input).");
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	num_changed
+}
+
+/// Reduce all redundant edges from the edge weight graph.
+///
+/// Returns the number of edges removed.
+///
+/// It is strictly assumed that the `who` attribute of all provided assignments are unique. The
+/// result will most likely be corrupt otherwise.
+///
+/// O(|Ew| ⋅ m)
+fn reduce_all<A: IdentifierT>(assignments: &mut Vec<StakedAssignment<A>>) -> u32 {
```

Comment:
This function is basically not understandable.

Reply:
So `reduce()` is the public function 
and it calls into `reduce_all()` and `reduce_4`. The former reduces cycles of all length. The former only length 4. I improved the doc.